File: src/main/java/komu/blunt/parser/PatternParser.java
package komu.blunt.parser;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.Lists;
import komu.blunt.types.patterns.Pattern;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static com.google.common.base.Preconditions.checkNotNull;
import static komu.blunt.parser.TokenType.*;
import static komu.blunt.types.DataTypeDefinitions.*;
import static komu.blunt.types.patterns.Pattern.*;

final class PatternParser {

    private final Lexer lexer;

    @SuppressWarnings("unchecked")
    private static final List<TokenType<?>> PATTERN_START_TOKENS =
        Arrays.asList(LPAREN, LBRACKET, LITERAL, IDENTIFIER, TYPE_OR_CTOR_NAME, UNDERSCORE);

    PatternParser(Lexer lexer) {
        this.lexer = checkNotNull(lexer);
    }

    // <literal> | <variable> | ( <pattern> ) | <constructor> <pattern>* |
    public Pattern parsePattern() {
        if (lexer.nextTokenIs(TYPE_OR_CTOR_NAME)) {
            String name = lexer.readTokenValue(TYPE_OR_CTOR_NAME);
            ImmutableList.Builder<Pattern> args = ImmutableList.builder();

            while (lexer.nextTokenIsOneOf(PATTERN_START_TOKENS))
                args.add(parseSimplePattern());

            return constructor(name, args.build());
        } else {
            return parseSimplePattern();
        }
    }

    public Pattern parseSimplePattern() {
        Pattern pattern = parsePrimitivePattern();

        if (lexer.nextTokenIs(OPERATOR) && lexer.peekTokenValue(OPERATOR).isConstructor()) {
            Operator op = lexer.readTokenValue(OPERATOR);

            pattern = constructor(op.toString(), pattern, parsePattern());
        }

        return pattern;
    }

    private Pattern parsePrimitivePattern() {
        if (lexer.nextTokenIs(LITERAL)) {
            return literal(lexer.readTokenValue(LITERAL));

        } else if (lexer.readMatchingToken(UNDERSCORE)) {
            return wildcard();

        } else if (lexer.nextTokenIs(IDENTIFIER)) {
            return variable(lexer.readTokenValue(IDENTIFIER));

        } else if (lexer.readMatchingToken(LPAREN)) {
            return parseParens();

        } else if (lexer.nextTokenIs(TYPE_OR_CTOR_NAME)) {
            return constructor(lexer.readTokenValue(TYPE_OR_CTOR_NAME));

        } else if (lexer.nextTokenIs(LBRACKET)) {
            return parseBrackets();


        } else {
            throw lexer.expectFailure("pattern");
        }
    }

    private Pattern parseBrackets() {
        lexer.expectToken(LBRACKET);

        if (lexer.readMatchingToken(RBRACKET))
            return constructor(NIL);
        
        List<Pattern> patterns = new ArrayList<>();

        patterns.add(parsePattern());
        while (lexer.readMatchingToken(COMMA))
            patterns.add(parsePattern());

        lexer.expectToken(RBRACKET);

        return createList(patterns);
    }

    private Pattern createList(List<Pattern> patterns) {
        Pattern result = constructor(NIL);

        for (Pattern pattern : Lists.reverse(patterns))
            result = constructor(CONS, pattern, result);

        return result;
    }

    private Pattern parseParens() {
        if (lexer.readMatchingToken(RPAREN))
            return constructor(UNIT);

        ImmutableList.Builder<Pattern> patterns = ImmutableList.builder();

        do {
            patterns.add(parsePattern());
        } while (lexer.readMatchingToken(COMMA));

        lexer.expectToken(RPAREN);

        ImmutableList<Pattern> pts = patterns.build();
        if (pts.size() == 1)
            return pts.get(0);
        else
            return constructor(tupleName(pts.size()), pts);
    }
}

File: src/main/kotlin/komu/blunt/parser/PatternParser.kt
package komu.blunt.parser

import komu.blunt.objects.Symbol
import komu.blunt.parser.TokenType.Companion.COMMA
import komu.blunt.parser.TokenType.Companion.IDENTIFIER
import komu.blunt.parser.TokenType.Companion.LBRACKET
import komu.blunt.parser.TokenType.Companion.LITERAL
import komu.blunt.parser.TokenType.Companion.LPAREN
import komu.blunt.parser.TokenType.Companion.RBRACKET
import komu.blunt.parser.TokenType.Companion.RPAREN
import komu.blunt.parser.TokenType.Companion.TYPE_OR_CTOR_NAME
import komu.blunt.types.ConstructorNames
import komu.blunt.types.patterns.Pattern
import java.util.*

final class PatternParser(val lexer: Lexer) {

    private val PATTERN_START_TOKENS =
        listOf(LPAREN, LBRACKET, LITERAL, IDENTIFIER, TYPE_OR_CTOR_NAME, TokenType.UNDERSCORE)

    // <literal> | <variable> | ( <pattern> ) | <constructor> <pattern>* |
    fun parsePattern(): Pattern {
        if (lexer.nextTokenIs(TYPE_OR_CTOR_NAME)) {
            val name = lexer.readTokenValue(TYPE_OR_CTOR_NAME)
            val args = ArrayList<Pattern>()

            while (lexer.nextTokenIsOneOf(PATTERN_START_TOKENS))
                args.add(parseSimplePattern())

            return Pattern.Constructor(name, args)
        } else {
            return parseSimplePattern()
        }
    }

    fun parseSimplePattern(): Pattern {
        var pattern = parsePrimitivePattern()

        if (lexer.nextTokenIs(TokenType.OPERATOR) && lexer.peekTokenValue(TokenType.OPERATOR).isConstructor) {
            val op = lexer.readTokenValue(TokenType.OPERATOR)

            pattern = Pattern.Constructor(op.toString(), pattern, parsePattern())
        }

        return pattern
    }

    private fun parsePrimitivePattern(): Pattern =
        when {
            lexer.nextTokenIs(LITERAL) ->
                Pattern.Literal(lexer.readTokenValue(LITERAL))
            lexer.readMatchingToken(TokenType.UNDERSCORE) ->
                Pattern.Wildcard
            lexer.nextTokenIs(IDENTIFIER) ->
                Pattern.Variable(Symbol(lexer.readTokenValue(IDENTIFIER)))
            lexer.nextTokenIs(LPAREN) ->
                parseParens()
            lexer.nextTokenIs(TYPE_OR_CTOR_NAME) ->
                Pattern.Constructor(lexer.readTokenValue(TYPE_OR_CTOR_NAME))
            lexer.nextTokenIs(LBRACKET) ->
                parseBrackets()
            else ->
                lexer.expectFailure("pattern")
        }

    private fun parseBrackets(): Pattern =
        lexer.inBrackets {
            if (lexer.nextTokenIs(RBRACKET))
                Pattern.Constructor(ConstructorNames.NIL)
            else
                createList(lexer.sepBy(TokenType.COMMA) { parsePattern() })
        }

    private fun createList(patterns: List<Pattern>): Pattern {
        var result = Pattern.Constructor(ConstructorNames.NIL)

        for (pattern in patterns.reversed())
            result = Pattern.Constructor(ConstructorNames.CONS, pattern, result)

        return result
    }

    private fun parseParens(): Pattern = lexer.inParens {
        if (lexer.nextTokenIs(RPAREN)) {
            Pattern.Constructor(ConstructorNames.UNIT)
        } else {
            val patterns = lexer.sepBy(COMMA) { parsePattern() }
            patterns.singleOrNull() ?: Pattern.Constructor(ConstructorNames.tupleName(patterns.size), patterns)
        }
    }
}
